docs/WritingADeviceDriver.md

```diff
@@ -0,0 +1,220 @@
+
+There is no requirement to expose a device as a single MMIO region.
+You may wish to define multiple regions, which can be as small as a single byte, so that you can privilege separate your device driver.
+
+Some devices have a very large control structure.
+For example, the platform-local interrupt controller is many KiBs.
+We don't define a C structure that covers every single field for this and instead just use `uint32_t` as the type for `MMIO_CAPABILITY`, which lets us treat the space as an array of 32-bit control registers.
+
+Handling interrupts
+-------------------
+
+To be able to handle interrupts, you must have a [software capability](SoftwareCapabilities.md) that [authorises access to the interrupt](Interrupts.md).
+For the ethernet device that we've been using as an example, you would typically request one with this macro invocation:
+
+```c
+DECLARE_AND_DEFINE_INTERRUPT_CAPABILITY(ethernetInterruptCapability, Ethernet, true, true);
+```
+
+If you wish to share this between multiple compilation units, you can use the separate `DECLARE_` and `DEFINE_` forms (see [`interrupt.h`](../sdk/include/interrupt.h)) but the combined form is normally most convenient.
+This macro takes four arguments:
+
+ 1. The name that we're going to use to refer to this capability.
+    The name `ethernetInterruptCapability` is arbitrary, you can use whatever makes sense to you.
+ 2. The name of the interrupt, from the board description file (`Ethernet`, in this case).
+ 3. Whether this capability authorises waiting for this interrupt (this will almost always be `true`).
+ 4. Whether this capability authorises acknowledging the interrupt so that it can fire again.
+    This will almost always be true in device drivers but should generally be true for only one compartment (for each interrupt), whereas multiple compartments may wish to observe interrupts for monitoring.
+
+As with the MMIO capabilities, compartments 
+
+```json
+        {
+          "contents": "10000101",
+          "kind": "SealedObject",
+          "sealing_type": {
+            "compartment": "sched",
+            "key": "InterruptKey",
+            "provided_by": "build/cheriot/cheriot/release/example-firmware.scheduler.compartment",
+            "symbol": "__export.sealing_type.sched.InterruptKey"
+          }
+```
+
+The sealing type tells you that this is an interrupt capability (it's sealed with the `InterruptKey` type, provided by the scheduler).
+The contents lets you audit what this authorises.
+The first two bytes are a 16-bit (little-endian on all currently supported targets) integer containing the interrupt number, so 1000 means 16 (our Ethernet interrupt number).
+The next two bytes are boolean values reflecting the last two arguments to the macro, so this authorises both waiting and clearing the macro.
+Again, this can form part of your firmware auditing.
```

Comment:
We really need tooling for this. Manually interpreting binary representation of structs is not fun.

Reply:
Yup, there are a bunch of auditing things that consume JSON, we should pick one and provide helper functions for it.